### PR TITLE
Processing Data in Models

### PR DESCRIPTION
**Pull request by Whoistolu**
# Processing Data in Models
## In this milestone. I;
-  Created model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
-  Set up associations between models.

### Added:

-  Users
    - A method that returns the 3 most recent posts for a given user.
-  Posts
    - A method that updates the posts counter for a user.
    - A method that returns the 5 most recent comments for a given post.
-  Comments
    - A method that updates the comments counter for a post.
-  Likes
    - A method that updates the likes counter for a post.

### Additionally:

 -  Created a method called current_user in your ApplicationController.
-  A current_user method that returns the first user from a database.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Post < ApplicationRecord
+  has_many :comments, foreign_key: 'post_id'
+  has_many :likes, foreign_key: 'post_id'
+  belongs_to :user
```
Comment (petudeveloper) [inline app/models/post.rb]: The foreign key is improperly implemented. As per the given ERD, the Post association should have Post belong to User through Author, an alias of User, like below:
```
belongs_to :author, class_name: 'User'
has_many :comments
has_many :likes
```

Kindly ensure to improve upon this aspect.

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
```
Comment (petudeveloper) [inline app/models/comment.rb]: Foreign key is improperly implemented. As per the given ERD, the Comment association should have Comment belong to User through Author, an alias of User, like below:
`belongs_to :author, class_name: 'User'`

Kindly ensure to improve upon this aspect.

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
```
Comment (petudeveloper) [inline app/models/like.rb]: The foreign key is improperly implemented. As per the given ERD, the Like association should have Like belong to User through Author, an alias of User, like below:
`belongs_to :author, class_name: 'User'`

Kindly ensure to improve upon this aspect.

File: app/models/post.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    comment.order('created_at Desc').limit(5)
+  end
+
+  after_save :update_posts_counter
+
+  private
+
+  def update_posts_counter
+    user.increment!(:posts_counter)
+  end
```
Comment (petudeveloper) [inline app/models/post.rb:16]: Great work here using [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to the database. 👏👏

File: app/models/comment.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+  after_save :update_comments_counter
+
+  private
+
+  def update_comments_counter
+    post.increment!(:comments_counter)
+  end
+end
```
Comment (petudeveloper) [inline app/models/comment.rb:12]: Great work here using [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_comments_counter` method every time a new post has been saved to the database. 👏👏

File: app/models/like.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+  after_save :update_comments_counter
+
+  private
+
+  def update_likes_counter
+    post.increment!(:likes_counter)
+  end
+end
```
Comment (petudeveloper) [inline app/models/like.rb:12]: Great work here using [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_likes_counter` method every time a new post has been saved to the database. 👏👏

**Review by petudeveloper — CHANGES_REQUESTED** [review]
Hi @Whoistolu, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

[OPTIONAL] You can consider adding instructions for creating and migrating the databases before running rails server:
```
rails db:create
rails db:migrate
```
This will make your project look more professional and will help the more innexperinece developers.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/models/post.rb (hunk 1 above)
Comment (Whoistolu) [reply app/models/post.rb]: Hi @petudeveloper Thank you for the feedback. I have fixed the foreign key implementation, however, I do not think I need to use `class_name: 'User'` as the user_id is the foreign key I am using and that will automatically match the User class

File: app/models/comment.rb (hunk 2 above)
Comment (Whoistolu) [reply app/models/comment.rb]: Hello @petudeveloper I do not think I need to use class_name: 'User' as the user_id is the foreign key I am using and that will automatically match the User class

File: app/models/like.rb (hunk 3 above)
Comment (Whoistolu) [reply app/models/like.rb]: Hi @petudeveloper I do not think I need to use class_name: 'User' as the user_id is the foreign key I am using and that will automatically match the User class

File: app/models/post.rb (hunk 4 above)
Comment (Whoistolu) [reply app/models/post.rb:16]: Hi @petudeveloper  I do not think I need to use class_name: 'User' as the user_id is the foreign key I am using and that will automatically match the User class

File: app/models/comment.rb (hunk 2 above)
Comment (arslanbisharat) [reply app/models/comment.rb]: @Whoistolu  Actually the Code Reviewer is correct. You should have` belongs_to :author, class_name: 'User'`. Let's go back to your [project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project.md).

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
# Changes Requested, 

Hi there, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have limited number of code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: app/models/comment.rb (hunk 2 above)
Comment (Whoistolu) [reply app/models/comment.rb]: @arslanbisharat I have implemented the recommended changes, although I was using `User` which has a `user_id` in the `schema` so rails will automatically figure that out. 
The project requirement never stated anything like `class_name`

**Review by Dipeshtwis — APPROVED** [review]
# Status: Approved ✔️

## Hi @Whoistolu, 

Good job on implementing the suggested changes. Well Done.
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

**You may close this Pull Request and merge the changes**

  
**Keep being awesome devs and good luck with your next milestone!**
---------

Cheers and Happy coding!👏👏👏
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
_**Dipesh Kumar**_

--------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project(check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Whoistolu** [comment]
@arslanbisharat I have implemented the recommended changes, although I was using `User` which has a `user_id` in the `schema` so rails will automatically figure that out.  
The project requirement never stated anything like `class_name`.

**Pull request merged**